### DISCUSSION TITLE
Lettuce publish doesn't work(disconnect) after some time.

### DISCUSSION THREAD

**Discussion by Obyvante** [redis/lettuce#2818] [Q&A]
#### Problem

I have a discord bot(discord-jda) that uses Lettuce Redis client. I use Lettuce client and connection with default settings(we could say everything is default).

After 2-5 minutes(or more), when a do "connection.sync().publish" it stucks one minute then throw RedisServerDisconnect error with or without timeout exception. The thing is I can set, del, etc. without any problem! It only happens when publishing something.

**Important:** It happens only if the Redis database server and application are in different region/network. If Redis database server and application are in same network/dedicated server, it works.

#### Base redis connection code
```java
    /**
     * Creates a new {@link BaseRedisProvider} instance for a database.
     * <p>
     * <b>If you are encountering a connection error about netty unix, please run your Java application with this flag; -Dio.lettuce.core.epoll=false</b>
     *
     * @param redisInformation {@link RedisInformation} Redis information.
     * @param codec            {@link RedisCodec} Codec for the database.
     * @throws NullPointerException     If {@param redisInformation} or {@param codec} is {@code null}.
     * @throws IllegalArgumentException If {@param port} is less than 0 or greater than 65535.
     * @throws IllegalArgumentException If {@param host} is empty.
     * @see Database#Database(String, int)
     */
    public BaseRedisProvider(@NotNull final RedisInformation redisInformation, @NotNull final RedisCodec<K, V> codec) {
        super(redisInformation.host(), redisInformation.port());
        this.codec = Objects.requireNonNull(codec, "Codec cannot be null.");

        // Client and connection initialization.
        this.client = RedisClient.create(RedisURI.builder()
                .withHost(this.host)
                .withPort(this.port)
                .withPassword(redisInformation.password().toCharArray())
                .build());

        // Base client connection.
        // One connection is enough for everything.
        this.connection = this.client.connect(codec);
        this.connection.addListener(BaseRedisProvider.STATE_HANDLER);

        // PubSub connection.
        // One connection is enough for everything.
        this.pubSubConnection = this.client.connectPubSub(codec);
        this.pubSubConnection.addListener(BaseRedisProvider.STATE_HANDLER);

        // Redis shutdown hook.
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            BaseRedisProvider.closing = true;
            if (this.connection.isOpen()) this.connection.close();
            if (this.pubSubConnection.isOpen()) this.pubSubConnection.close();
        }, "redis-shutdown-hook"));
    }
```

#### Test code
```java
this.connection.sync().set("test", "value");
this.connection.sync().get("test");
this.connection.sync().publish("test-channel", "test"); // <-- Error 4-5 minutes after application start. Otherwise, no error with even 1k+ execution of this code.
```

#### Environment

- Lettuce version(s): lettuce-core-6.3.0.RELEASE
- Redis version: 7.2.1

**Comment by Obyvante** [comment]
**New information**
It happens only if the Redis database server and application are in different region/network. If Redis database server and application are in same network/dedicated server, it works.

**Comment by tishun** [comment]
I see that in the test code you use the `connection` object for both pubsub and non-pubsub operations, but in the connection code you have two different connection objects - `connection` and `pubSubConnection`.

Does the operation fail if you also use the `pubSubConnection`? E.g. 


```java
this.connection.sync().set("test", "value");
this.connection.sync().get("test");
this.pubSubConnection.sync().publish("test-channel", "test"); // THIS LINE IS ALTERED
```

**Reply by Obyvante** [reply]
I am so sorry for the late response(it was due to a mail notification problem).

I have tried both of them. It only happens when the database(redis) and the application server are on different networks. If I host both of them on the same networks, everything works perfectly. I have two other databases; Postgresql and MongoDB. They are working normally. I do believe something is wrong with the pub/sub-models.

```java
this.connection.sync().set("test", "value"); <- works
this.connection.sync().get("test"); <- works
this.pubSubConnection.sync().publish("test-channel", "test"); <- doesn't work
```

Thank you for your answer!

**Reply by tishun** [reply]
I was not really able to reproduce this locally. Could you help with some sample set up?
When you say "different networks" - what does that specifically mean? Do you use some docker compose / k8s setup?
If you could provide an environment for test or at least the steps to create one this would greatly help debug the issue.
Thanks!
